### PR TITLE
fix: fix serialisation of arithmetic expressions

### PR DESCRIPTION
**Pull request by guipublic**
Fix an issue related to the infamous "linear eval is no 0" of BB TP.
The witness of the BB polygates were sometimes overwritten during serialisation.
n.b could not fully test this, but I added a unit test.

### CODE REVIEW

File: common/src/barretenberg_structures.rs [hunk 1]
```diff
@@ -91,6 +106,21 @@ impl Constraint {
 
         buffer
     }
+
+    fn set_linear_term(&mut self, coef: Scalar, witness: i32) {
+        if self.a == 0 || self.a == witness {
+            self.a = witness;
+            self.ql = coef;
```
Comment (TomAFrench) [inline common/src/barretenberg_structures.rs:113]: spellcheck on `coef`

File: common/src/barretenberg_structures.rs [hunk 2]
```diff
@@ -938,87 +968,81 @@ impl From<&Circuit> for ConstraintSystem {
...
+        let x1 = Witness::new(1);
+        let x3 = Witness::new(3);
+        let two = Scalar::one() + Scalar::one();
+        let e = Expression {
+            mul_terms: vec![(Scalar::one(), x1, x1)],
+            linear_combinations: vec![(two, x1), (-Scalar::one(), x3)],
+            q_c: Scalar::one(),
+        };
+        let constrain = serialize_arithmetic_gates(&e);
+        assert!(constrain.a == 1);
```
Comment (TomAFrench) [inline common/src/barretenberg_structures.rs:1012]:
```suggestion
        assert_eq!(constrain.a, 1);
```

and all other instances.

File: common/src/barretenberg_structures.rs [hunk 3]
```diff
@@ -74,6 +74,21 @@ pub struct Constraint {
...
+            b: 0,
+            c: 0,
+            qm: Scalar::zero(),
+            ql: Scalar::zero(),
+            qr: Scalar::zero(),
+            qo: Scalar::zero(),
+            qc: Scalar::zero(),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (guipublic) [reply common/src/barretenberg_structures.rs:90]: No, else I would have.

File: common/src/barretenberg_structures.rs (hunk 1 above)
Comment (guipublic) [reply common/src/barretenberg_structures.rs:113]: ?

Comment (TomAFrench) [reply common/src/barretenberg_structures.rs:113]: This is throwing warnings in CI for the spellchecker.

**Comment by kevaundray** [comment]
@guipublic bumping -- can you fix merge conflicts?

File: common/src/barretenberg_structures.rs (hunk 2 above)
Comment (guipublic) [reply common/src/barretenberg_structures.rs]: done

File: common/src/barretenberg_structures.rs (hunk 1 above)
Comment (guipublic) [reply common/src/barretenberg_structures.rs]: Done

**Comment by guipublic** [comment]
> @guipublic bumping -- can you fix merge conflicts?

I did, but build is now failing due to nix errors??

**Comment by TomAFrench** [comment]
This is failing due to unresolved merge conflicts. It's still trying to use the `Scalar` type alias which no longer exists.

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**